### ISSUE TITLE
Support for RInChI

### ISSUE THREAD

**Issue by FelixBaensch** [cdk/cdk#1063]
I wonder if there is already a way to create a reaction InChI or am I just not finding it.
Or is it planned to integrate this in future, since RInChI is part of the JNA InChI?

**Comment by johnmay** [comment]
We had a patch a while ago but it made more sense to push that down to JNA-Inchi - https://github.com/dan2097/jna-inchi. Now it is in there we should add the RInChI support.

**Comment by FelixBaensch** [comment]
That sounds good, thank you. Can you say when this will be tackled? CDK v3.0?

**Comment by johnmay** [comment]
No, but we are always open to contributions.

**Comment by uli-f** [comment]
@FelixBaensch I have a patch available that is fully functional and builds on the API available at https://github.com/dan2097/jna-inchi. Adding this ability to jna inchi was sponsored by the company I work for. However, there was a concern about the size increase of CDK that results from including the RInChI binaries of jna-inchi in addition to the inchi binaries. Maybe we can discuss again with CDK devs. Feel free to contact me if you want to have a chat about this.

**Comment by johnmay** [comment]
This is somewhat InChI's problem - when I last looked the RInChI code was out of sync with the main parts. It's a few lines of code to create a valid and correct RInChI from the underlying InChI's - maybe a bit more effort for all the INCHI-KEY variants.

The argument to add to JNA InChI is it tracks the "official" algorithm but currently that really doesn't do much and is akin to the famous "isNumber" JS package. https://www.youtube.com/watch?v=8XF1qv5qfZM

**Comment by johnmay** [comment]
To clarify if the main InChI lib just did the RInChI bits it would be easier to swallow.

**Comment by uli-f** [comment]
> To clarify if the main InChI lib just did the RInChI bits it would be easier to swallow.

Agreed.

There is an [issue](https://github.com/IUPAC-InChI/RInChI/issues/7) in the RInChI repo that addresses the ability _to not link any InChI code into RInChI_ and have _the end user of RInChI [...] provide already-calculated component InChI-s to RInChI_, but it seems to have gone stale, so probably not worth it getting our hopes up too much.

**Comment by uli-f** [comment]
> If I understand you correctly, and please correct me if I am wrong, would you prefer a Java-native implementation of the RInChI code based on the existing InChI stuff directly into the CDK?

Imo this is the next best alternative.

@FelixBaensch I am pretty sure that there are a few things we can re-use from the existing code that is based on the rinchi implementation in jna-inchi. At the very least most of the testing code might be easy enough to re-purpose for a Java native approach. Yes, please let's get in touch and see how we can approach this together.

**Comment by fbaensch-beilstein** [comment]
> There is an [issue](https://github.com/IUPAC-InChI/RInChI/issues/7) in the RInChI repo that addresses the ability _to not link any InChI code into RInChI_ and have _the end user of RInChI [...] provide already-calculated component InChI-s to RInChI_, but it seems to have gone stale, so probably not worth it getting our hopes up too much.

I ask Gerd if this will be dealt with further.

**Comment by johnmay** [comment]
>> If I understand you correctly, and please correct me if I am wrong, would you prefer a Java-native implementation of the RInChI code based on the existing InChI stuff directly into the CDK?

> Imo this is the next best alternative.

Correct... however I propssed this when Niko (I think) and he said we would rather use the "official" build.

```
package com.nextmovesoftware.pistachio;

import com.nextmovesoftware.pistachio.data.Component;
import com.nextmovesoftware.pistachio.data.Reaction;
import org.openscience.cdk.exception.CDKException;
import org.openscience.cdk.inchi.InChIGeneratorFactory;
import org.openscience.cdk.interfaces.IAtomContainer;
import org.openscience.cdk.silent.SilentChemObjectBuilder;
import org.openscience.cdk.smiles.SmilesParser;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class RInChI {

  private static final String RINCHI_VERSION = "1.00";
  private static final String INCHI_VERSION  = "1S";
  private static final Object lock           = new Object();

  private static String getInChI(String smi)
  {
    try {
      SmilesParser   smipar = new SmilesParser(SilentChemObjectBuilder.getInstance());
      IAtomContainer mol    = smipar.parseSmiles(smi);
      // InChI was not thread safe pre 1.0.5
      synchronized (lock) {
        return InChIGeneratorFactory.getInstance().getInChIGenerator(mol).getInchi();
      }
    } catch (CDKException e) {
      return null;
    }
  }

  public static String getRInChI(Reaction reaction)
  {

    List<String> reactants = new ArrayList<>();
    List<String> products  = new ArrayList<>();
    List<String> agents    = new ArrayList<>();

    for (Component c : reaction.getComponents()) {

      String inchi = c.getInChI();

      if (inchi == null && c.getSmiles() != null) {
        inchi = getInChI(c.getSmiles());
      }

      if (inchi != null && inchi.startsWith("InChI=1S/")) {
        inchi = inchi.substring("InChI=1S/".length());
        switch (c.getRole()) {
          case Reactant:
            reactants.add(inchi);
            break;
          case Product:
            products.add(inchi);
            break;
          default:
            agents.add(inchi);
            break;
        }
      }
    }

    if (reactants.isEmpty() || products.isEmpty())
      return null;

    Collections.sort(reactants);
    Collections.sort(agents);
    Collections.sort(products);

    String reactantsPart = String.join("!", reactants);
    String productsPart  = String.join("!", products);
    String agentsPart    = String.join("!", agents);

    StringBuilder sb = new StringBuilder();
    sb.append("RInChI=")
      .append(RINCHI_VERSION)
      .append(".")
      .append(INCHI_VERSION)
      .append("/");

    // direction forwards?
    if (reactantsPart.compareTo(productsPart) <= 0) {
      sb.append(reactantsPart);
      sb.append("<>");
      sb.append(productsPart);
      if (!agentsPart.isEmpty()) {
        sb.append("<>");
        sb.append(agentsPart);
      }
      sb.append("/d+");
    } else {  // ... direction => backwards
      sb.append(productsPart);
      sb.append("<>");
      sb.append(reactantsPart);
      if (!agentsPart.isEmpty()) {
        sb.append("<>");
        sb.append(agentsPart);
      }
      sb.append("/d-");
    }
    return sb.toString();
  }
}
```

**Comment by johnmay** [comment]
That is the crux of it... no RINCHI-KEYS but not hard to add.

**Comment by uli-f** [comment]
Thanks @johnmay for the code outline. @FelixBaensch and I will work on a PR using the public facing API available for generating InChIs in CDK as a template for the RInChI API. Happy to get your feedback once we have a draft in place. If there is anything you would like to see please feel invited to let us know.

**Comment by johnmay** [comment]
Closed via #1137
